common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java

```diff
@@ -383,26 +384,29 @@ private class ManagedBufferIterator implements Iterator<ManagedBuffer> {
       String appId = msg.appId;
       String execId = msg.execId;
       String[] blockIds = msg.blockIds;
+      size = blockIds.length;
       String[] blockId0Parts = blockIds[0].split("_");
-      if (blockId0Parts.length == 4 && blockId0Parts[0].equals(SHUFFLE_BLOCK_ID)) {
-        final int shuffleId = Integer.parseInt(blockId0Parts[1]);
-        final int[] mapIdAndReduceIds = shuffleMapIdAndReduceIds(blockIds, shuffleId);
-        size = mapIdAndReduceIds.length;
-        blockDataForIndexFn = index -> blockManager.getBlockData(appId, execId, shuffleId,
-          mapIdAndReduceIds[index], mapIdAndReduceIds[index + 1]);
-      } else if (blockId0Parts.length == 5 && blockId0Parts[0].equals(SHUFFLE_CHUNK_ID)) {
+      if (isShuffleBlock(blockIds[0])) {
+        blockDataForIndexFn = index -> {
+          if (!isShuffleBlock(blockIds[index])) {
+            throw new IllegalArgumentException("Unexpected block id format: " + blockIds[index]);
+          }
+          return blockManager.getBlockData(appId, execId, blockIds[index]);
+        };
+      } else if (blockId0Parts.length == 5 && isShuffleChunk(blockIds[0])) {
         final int shuffleId = Integer.parseInt(blockId0Parts[1]);
         final int shuffleMergeId = Integer.parseInt(blockId0Parts[2]);
         final int[] reduceIdAndChunkIds = shuffleReduceIdAndChunkIds(blockIds, shuffleId,
           shuffleMergeId);
-        size = reduceIdAndChunkIds.length;
         blockDataForIndexFn = index -> mergeManager.getMergedBlockData(msg.appId, shuffleId,
-          shuffleMergeId, reduceIdAndChunkIds[index], reduceIdAndChunkIds[index + 1]);
-      } else if (blockId0Parts.length == 3 && blockId0Parts[0].equals("rdd")) {
-        final int[] rddAndSplitIds = rddAndSplitIds(blockIds);
-        size = rddAndSplitIds.length;
-        blockDataForIndexFn = index -> blockManager.getRddBlockData(appId, execId,
-          rddAndSplitIds[index], rddAndSplitIds[index + 1]);
+          shuffleMergeId, reduceIdAndChunkIds[2 * index], reduceIdAndChunkIds[2 * index + 1]);
```

Comment:
Based on a quick look, this  should be `reduceIdAndChunkIds[index], reduceIdAndChunkIds[index + 1]` instead ?